qutip/scattering.py

```diff
@@ -0,0 +1,398 @@
+    waveguides.
+
+    Parameters
+    ----------
+    evolver : Evolver
+        Evolver-wrapped Hamiltonian describing the system
+    c_ops : list
+        list of collapse operators for each waveguide; these are assumed to
+        include spontaneous decay rates, e.g.
+        :math:`\\sigma = \\sqrt \\gamma \\cdot a`
+    taus_list : list-like
+        list of (list of emission times) for each waveguide
+
+    Returns
+    -------
+    Omega : Qobj
+        The temporal scattering operator with dimensionality equal to the
+        system state
+    """
+
+    omega = 1
+
+    # Extract the full list of taus
+    taus = [(0.0, -1)]  # temporal "ground state" for arbitrary waveguide
+    for i, tauWG in enumerate(taus_list):
+        for tau in tauWG:
+            taus.append((tau, i))
+    taus.sort(key = lambda tup: tup[0])  # sort taus by time
+
+    # Compute Prod Ueff(tq, tq-1)
+    for i in range(1, len(taus)):
+        tq, q = taus[i]
+        tprev, _ = taus[i - 1]
+        omega = c_ops[q] * evolver.prop(tq, tprev) * omega
+
+    # Add the <0|Uff(TP, tm)|0> term
+    tmax = evolver.tlist[-1]
+    taumax, _ = taus[-1]
+    # if taus[-1] < tmax:
+    omega = evolver.prop(tmax, taumax) * omega
+
+    return omega
+
+
+def temporal_basis_vector(waveguide_emission_indices, n_time_bins):
+    """
+    Generate a temporal basis vector for emissions at specified time bins into
+    specified waveguides
+
+    Parameters
+    ----------
+    waveguide_emission_indices : list or tuple
+        list of indices where photon emission occurs for each waveguide,
+        e.g. [[t_1]_wg1, [t_1, t_2]_wg2, []_wg3, [t_1, t_2, t_3]_wg4]
+    n_time_bins : int
+        number of time bins; the range over which each index can vary
+
+    Returns
+    -------
+    temporal_basis_vector : Qobj
+        A basis vector representing photon scattering at the specified indices.
+        If there are W waveguides, T times, and N photon emissions, then the
+        basis vector has dimensionality (W*T)^N.
+    """
+
+    # Cast waveguide_emission_indices to list for mutability
+    waveguide_emission_indices = [list(i) for i in waveguide_emission_indices]
+
+    # Calculate total number of waveguides
+    W = len(waveguide_emission_indices)
+
+    # Calculate total number of emissions
+    num_emissions = sum([len(waveguide_indices) for waveguide_indices in
+                         waveguide_emission_indices])
+    if num_emissions == 0:
+        return basis(W * n_time_bins, 0)
+
+    # Pad the emission indices with zeros
+    offset_indices = []
+    for i, wg_indices in enumerate(waveguide_emission_indices):
+        offset_indices += [index + i * n_time_bins for index in wg_indices]
+
+    # Return an appropriate tensor product state
+    return tensor([basis(n_time_bins * W, i) for i in offset_indices])
+
+
+def temporal_scattered_state(H, psi0, n_emissions, c_ops, tlist,
+                             system_zero_state = None,
```

Comment:
`system_zero_state` -> `zero_state`: I picked this name to avoid confusion between the zero-state of the system and the vacuum state of the waveguide when referring to @kafischer 's paper.

`construct_effective_hamiltonian` -> `effective_hamiltonian`: I agree that this kwarg is long, but this suggested change would add a lot of ambiguity. (Does "effective_hamiltonian" mean that the provided hamiltonian is already in H_eff form? Or does it mean that the function should construct an effective hamiltonian from one that does not yet have the sum(-i/2 a^dag a) terms in it?) However, I'll try to think of a shorter name for this kwarg.